File: aoc24-solutions/day02/day02p02.kt
import kotlin.math.abs

fun is_safe(nums: List<Int>): Boolean {
    val asc = nums.windowed(2).all { (first, second) ->
        first < second && abs(first - second) in 1..3 
    }
    val desc = nums.windowed(2).all { (first, second) ->
        first > second && abs(first - second) in 1..3 
    }
    return asc || desc
}

fun main() {
    val result = generateSequence(::readLine).count { line ->
        val nums = line.split("\\s+".toRegex()).map { it.toInt() }
        nums.indices.any {
            is_safe(nums.take(it) + nums.drop(it + 1))
        }
    }
    println(result)
}

File: aoc24-solutions/day02/day02p02.py
import sys
from day02p01 import is_safe

def main():
    ans = 0
    for line in sys.stdin:
        nums = list(map(int, line.split()))
        for i in range(len(nums)):
            if (is_safe(nums[:i] + nums[i+1:])):
                ans += 1
                break
    return ans

if __name__ == "__main__":
    print(main())
